pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -125,15 +125,23 @@ func (c *clusterMemberRemovalController) attemptToScaleDown(ctx context.Context,
 	if desiredControlPlaneReplicasCount == 0 {
 		return fmt.Errorf("desired control plane replicas count cannot be empty")
 	}
-	if currentVotingMemberIPListSet.Len() <= desiredControlPlaneReplicasCount {
-		klog.V(4).Infof("Ignoring scale-down since the number of etcd voting members (%d) < desired number of control-plane replicas (%d) ", currentVotingMemberIPListSet.Len(), desiredControlPlaneReplicasCount)
+
+	liveVotingMembers, err := c.getAllVotingMembers(ctx)
+	if err != nil {
+		// TODO : update status condition
+		// TODO : should go degraded ?
+		return fmt.Errorf("could not list etcd members: %w", err)
+	}
+
+	if len(liveVotingMembers) < desiredControlPlaneReplicasCount {
+		klog.V(2).Infof("Ignoring scale-down since the number of etcd members (%d) < desired number of control-plane replicas (%d) ", len(liveVotingMembers), desiredControlPlaneReplicasCount)
 		return nil
 	}
 
 	// machines with master role and deletion hook
 	memberMachines, err := ceohelpers.CurrentMemberMachinesWithDeletionHooks(c.masterMachineSelector, c.masterMachineLister)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not find master machines with deletion hook: %w", err)
 	}
 
 	var votingMembersMachines []*machinev1beta1.Machine
```

Comment:
just to make the below logic consistent, should the `currentVotingMemberIPListSet` sourced from `liveVotingMembers` too? Otherwise you have one set generated by configmap, the other from the etcd memberlist.

Reply:
I actually wanted to do this, but did not want to change the logic completely. 

I had to use `liveVotingMembers` for deciding whether attempt scaling down, or skip. During testing, I noticed that the  configmap takes time to reflect the cluster membership, and the CEO keep spinning waiting for a new CM to be created. 

+1 to replace `currentVotingMemberIPListSet` with `liveVotingMembers`, wdyt @hasbro17  ?

Reply:
I had the opposite thought actually. For the default case when we're checking if scale-down is allowed i.e `votingMembers >= desiredControlPlaneReplicasCount` we don't want to make live calls to the etcd cluster.

The reason why we use `currentVotingMemberIPListSet` (sourced from the etcd-endpoints configmap) is so we're not inducing unnecessary load in the default case by using a cache instead of live calls to cluster. 

>During testing, I noticed that the configmap takes time to reflect the cluster membership, and the CEO keep spinning waiting for a new CM to be created.

Not sure what CM here is (assuming machine) but other than the delay @Elbehery Is there a reason why you can't use the `currentVotingMemberIPListSet` in L128-135 above? How much is the delay in practice? I would expect the etcd-endpoints controller to update within the minute once there is an actual membership change.

Reply:
Thanks @hasbro17  for your review 

> votingMembers >= desiredControlPlaneReplicasCount we don't want to make live calls to the etcd cluster.

the etcd-endpoint CM contains all the voting member ( i.e. non-learner ) whether `healthy` or `unhealthy`.  In https://github.com/openshift/cluster-etcd-operator/pull/937#discussion_r991786802 we agreed to constrain to the scaling-down invariant around `healthyVotingMembers` iiuc 

> we're not inducing unnecessary load in the default case by using a cache instead of live calls to cluster.

I think listing the cluster members is not a load, moreover the recent changes to `etcdcli` package reuses the connections :) 

> I would expect the etcd-endpoints controller to update within the minute once there is an actual membership change.

Yes this is correct, it takes around a minutes to update the CM .. I can change it back, but may I ask listing the healthy members directly from the etcd cluster is not correct ?

Reply:
> The reason why we use currentVotingMemberIPListSet (sourced from the etcd-endpoints configmap) is so we're not inducing unnecessary load in the default case by using a cache instead of live calls to cluster.

That makes sense, I think we make lots of MemberList calls across the operator already (health probe, etc). The connection to etcd is already open as Mustafa said, so getting the current list of members is not an expensive call.  

I'm all up for consolidating this listing into a single controller that queries this in the background every 5-10s and all other controllers just feed from that. 

> How much is the delay in practice?

I don't have a direct answer for that, but let's go with the case I've added to the quorum guard here: 
https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/etcdendpointscontroller/etcdendpointscontroller.go#L145

So when you're down to two nodes, we'll never update the CM until you add a new node. I realize now that this is stupid, but we have no other way to selectively tell the static pod controller not to rollout a revision based on that.

Reply:
> I'm all up for consolidating this listing into a single controller that queries this in the background every 5-10s and all other controllers just feed from that.

+1

>  we'll never update the CM until you add a new node.

In this case I think listing the members from etcd client makes more sense, or ?

Reply:
Yeah I forgot that we'll have client connections open from the client cache so I guess having this controller make periodic live list calls might not add a significant load. So I guess we can keep the live call here.

> So when you're down to two nodes, we'll never update the CM until you add a new node

Yeah that's not ideal especially since the apiserver is using the endpoints configmap and will end up talking to a non-existent endpoint if we have 2 members and the CM is outdated with 3. But it's a necessary trade-off that we've made given that we don't want quorum loss mid rollout.
Plus with CPMS once we should have a new machine who's member would soon be added and promoted right after the unhealthy one is removed so this gap shouldn't be too large.

Let's keep the live list for now, and leave off the consolidation of list calls to a better internal cache than the etcd-endpoints CM as a follow up.

Reply:
if we agree on sticking with the member listing, then I would definitely like to see `currentVotingMemberIPListSet` to come from the member listing. Otherwise this is going to be wildly inconsistent.